src/client/opamConfigCommand.ml

```diff
@@ -217,6 +217,25 @@ let print_eval_env ~csh ~sexp ~fish env =
   else
     print_env env
 
+let ensure_env_aux ?(set_opamroot=false) ?(set_opamswitch=false) ?(force_path=true) gt switch =
+  let env_file = OpamPath.Switch.environment gt.root switch in
+  if not (OpamFile.exists env_file) then
+    Some (OpamSwitchState.with_ `Lock_none gt @@ fun st ->
```

Comment:
A good reason to not take the write lock directly in this case ? Lock upgrades should be avoided unless absolutely necessary (and the scenarios for deadlocks have been considered)

Reply:
Happy to change this, but this is a refactoring - it's exactly the code deleted below.

Reply:
Oh right! I should ask @rjbou then :)

Reply:
`env` file is only regenerated & written if a write lock can be taken (not `--safe`, cf. 3691). We don't want to block `opam env` because of the missing environment file.

Reply:
Yes, was I was suggesting is stg along the lines of:
```
  if not safe_mode && not OpamFile.exists env_file then OpamSwitchState.with_ `Lock_write gt @@ ... (* regenerate *)
  else OpamSwitchState.with_ `Lock_none gt @@ ...
```

the point would just be to avoid the use of `OpamSwitchState.with_write_lock` really, since we could detect the need beforehand and load with a write lock directly, which is always better & safer than upgrading existing locks.